### DISCUSSION TITLE
External api for async inital state in NextJs14, while using createStore with contextAPI.

### DISCUSSION THREAD

**Discussion by strizer9** [pmndrs/zustand#2488] [Q&A] [answered]
I want to setup my store with some APIs, in NextJs 14. 
Following this for nextJs setup - https://docs.pmnd.rs/zustand/guides/nextjs 
Similar discussion - https://github.com/pmndrs/zustand/discussions/676
_edited_ - stackblitz link - https://stackblitz.com/edit/nextjs-spgsd2?file=stores%2Fcounter-store.ts
Below code isnt working, I have tried  something like in counter-store-provider.tsx file -
`getData().then((data) => useStore.setState({ data.id}))`
 
```javascript
// src/stores/counter-store.ts
import { createStore } from 'zustand/vanilla'

export type CounterState = {
  count: number
  loading: boolean
}

export type CounterActions = {
  decrementCount: () => void
  incrementCount: () => void
}

export type CounterStore = CounterState & CounterActions

export const initCounterStore = async (): Promise<CounterState> => {


  try {
    const response = await fetch('https://jsonplaceholder.typicode.com/todos/1');
    const data = await response.json();
    console.log(data)
    return {
      count: data.id,
      loading: false,
    };
  } catch (error) {
    console.error('Error fetching data:', error);
    return defaultInitState
  }
}

export const defaultInitState: CounterState = {
  count: 10,
  loading: true,
}

export const createCounterStore = (
  initState: CounterState = defaultInitState,
) => {
  return createStore<CounterStore>()((set) => ({
    ...initState,
    decrementCount: () => set((state) => ({ count: state.count - 1 })),
    incrementCount: () => set((state) => ({ count: state.count + 1 })),
  }))
}
```

**Comment by dai-shi** [comment]
How is `initCounterStore` function used? Do you use it on the client only?

**Reply by strizer9** [reply]
As per this guide -  https://docs.pmnd.rs/zustand/guides/nextjs#initializing-the-store
Yes. 

In provider file, which is a client file, we are passing  `initCounterStore` in `createCounterStore(initCounterStore())`

**Comment by dbritto-dev** [comment]
@strizer9 would you mind sharing us a minimal repro on stackblitz?

**Reply by strizer9** [reply]
https://stackblitz.com/edit/nextjs-spgsd2?file=stores%2Fcounter-store.ts

Please refer this. 
The end goal is to initial store through an API.

**Reply by dbritto-dev** [reply]
@strizer9 you should follow Daishi's suggestion. The `initCounterStore` can't be async.

**Reply by dai-shi** [reply, marked as answer]
How about something like this? https://stackblitz.com/edit/nextjs-klyfxf?file=stores%2Fcounter-store.ts

**Reply by dai-shi** [reply]
#676 is a similar discussion.

**Reply by strizer9** [reply]
Thank you, it works. :)
